### PR TITLE
Enhanced RDoc for case mapping

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Adds file  doc/case_mapping.rdoc, which describes case mapping and provides a link target that methods doc can link to.

Revises:
- String#capitalize
- String#capitalize!
- String#casecmp
- String#casecmp?
- String#downcase
- String#downcase!
- String#swapcase
- String#swapcase!
- String#upcase
- String#upcase!
- Symbol#capitalize
- Symbol#casecmp
- Symbol#casecmp?
- Symbol#downcase
- Symbol#swapcase
- Symbol#upcase

### CODE REVIEW

**Comment by BurdetteLamar** [comment]
@duerst and @jeremyevans, I don't think the `:lithuanian` option does anything.  Should it be removed?

**Comment by duerst** [comment]
> @duerst and @jeremyevans, I don't think the `:lithuanian` option does anything. Should it be removed?

You're right that :lithuanian doesn't do anything currently. But it's allowed as an option, whereas almost all other symbols (such as :foo, :bar,...) are not allowed.

**Comment by BurdetteLamar** [comment]
@duerst, will you be reviewing this PR?

File: string.c [hunk 1]
```diff
@@ -7185,15 +7201,18 @@ rb_str_upcase_bang(int argc, VALUE *argv, VALUE str)
...
 /*
  *  call-seq:
- *     str.upcase              -> new_str
- *     str.upcase([options])   -> new_str
+ *    upcase(*options) -> string
+ *
+ *  Returns a string containing the upcased characters in +self+:
+ *
+ *     s = 'Hello World!' # => "Hello World!"
+ *     s.downcase         # => "hello world!"
```
Comment (jeremyevans) [inline string.c:7209]: This example should use `upcase`, not `downcase`

File: string.c [hunk 2]
```diff
@@ -7151,13 +7159,21 @@ upcase_single(VALUE str)
 
 /*
  *  call-seq:
- *     str.upcase!              -> str or nil
- *     str.upcase!([options])   -> str or nil
+ *    upcase!(*options) -> self of nil
```
Comment (jeremyevans) [inline string.c:7162]: `self of nil` -> `self or nil`

File: string.c [hunk 3]
```diff
@@ -7390,17 +7385,20 @@ rb_str_capitalize_bang(int argc, VALUE *argv, VALUE str)
...
  *  call-seq:
- *     str.capitalize              -> new_str
- *     str.capitalize([options])   -> new_str
+ *    capitalize(*options) -> string
  *
- *  Returns a copy of <i>str</i> with the first character converted to uppercase
- *  and the remainder to lowercase.
+ *  Returns a string containing the characters in +self+;
+ *  the first character is upcased;
+ *  the remaining characters are down cased:
```
Comment (jeremyevans) [inline string.c:7392]: `down cased` => `downcased`

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks good. Only a few minor changes requested, which don't require my further review.  So if @duerst is OK with this or doesn't want to review, it can be committed after the changes are made.

File: string.c (hunk 1 above)
Comment (BurdetteLamar) [reply string.c]: Done.

File: string.c (hunk 2 above)
Comment (BurdetteLamar) [reply string.c]: Done.

File: string.c (hunk 3 above)
Comment (BurdetteLamar) [reply string.c]: Done.

**Comment by BurdetteLamar** [comment]
Over to you, @duerst.

**Comment by duerst** [comment]
Thanks for all your work.

Sorry, I'm not sure how to review the changes the same way as @jeremyevans. I just looked at https://github.com/ruby/ruby/pull/5245/files, and am commenting on that.

https://www.unicode.org/reports/tr21/tr21-5.html is outdated, please don't reference it (I didn't even know this once existed!).

"This support will be extended to other encodings.": I may have written that somewhere, but it's vaporware, so please remove it.

"In most cases, the upper and lower cases for a character are the same size." -> "In most cases, case conversions of a string have the same number of characters."

"(see +:fold+ below)" -> "(see also +:fold+ below)"

"(see +:turkic+ below)" -> "(see also +:turkic+ below)"

"=== User-Specified Case Mapping" can be misunderstood to allow completely user-specified mappings. I'd change it to "=== Options for Case Mapping" or some such.

"Letters-only mapping:": This is confusing. Is 'Ü' not a letter? To everybody in Germany, it will be. -> "Mapping of ASCII characters only"

The following text mentions uppercase -> lowercase, but then the example uses upcase. It should be clear that :ascii applies to all 'directions'.

https://en.wikipedia.org/wiki/List_of_Turkic_languages is misleading, because for example Uzbeh is a Turkic language, but https://en.wikipedia.org/wiki/Uzbek_language#Writing_systems shows that they don't separate dotted/dotless I.

So I would write something like:  Full Unicode case mapping, adapted for the Turkic languages that distinguish dotted and dotless I, for example Turkish and Azeri.

For :lithuanian, just write "not yet implemented" and leave out the examples.

**Comment by BurdetteLamar** [comment]
@duerst, please comment or approve.

**Comment by BurdetteLamar** [comment]
I think I've addressed all review issues here.  If no one objects, I'll merge this.

**Comment by duerst** [comment]
I have looked at the changes that were done based on my comments. As far as I have seen, they should be okay.

**Review by duerst — APPROVED** [review]

**Pull request merged**